src/dune_util/located_error.mli

```diff
@@ -0,0 +1,14 @@
+(** A [Located_error.t] is morally a [User_error] annotated with an associated
+    directory. Namely, if a build action raises an error, we want to preserve
+    the directory that action was raised from. *)
+
+open Stdune
+
+exception E of User_message.t * Path.t
```

Comment:
This exception subtly changes how we handle user errors elsewhere in the code. See `expander.ml` where we treat user errors in a special way. More generally, it just seems fragile to me to extend the concept of a user error this way. Every time we add a new type of user error, we need to scan the entire code base to make sure we're consistently handling it without help from the type system.

I think we should consider alternatives such as:

- Optional dir argument in `User_error.E`.
- Use open types for the payload of user errors. E.g.

```ocaml
type error = ..
type error += User_error of Pp.t
type error += Located_error of Pp.t * Path.t
exception User_error of error
```

Reply:
I agree that it's pretty fragile to extend it this way. Personally, I'd rather just have a closed-world of errors (instead of the open variant or adding new exceptions), so the compiler can tell us where we're missing handling.

Directly referencing `Path.t` in `User_error` seems difficult, because `stdune-unstable/path.ml` needs to raise these errors in several cases, and from a brief scan it doesn't seem like it's going to be easy to split out a `Path0` type module either.